### PR TITLE
AG-7553 - Support switching chart type without invalidating chart references

### PR DESCRIPTION
**Pull request by alantreadway**
https://ag-grid.atlassian.net/browse/AG-7553

Adds support in Standalone Charts for switching chart type (cartesian vs. polar vs. hierarchy) without invalidating the `AgChartInstance` returned to user code.

Updates Integrated Charts to retain a reference to `AgChartInstance` across chart-type changes.

### CODE REVIEW

File: enterprise-modules/charts/src/charts/chartComp/chartProxies/chartProxy.ts [hunk 1]
```diff
@@ -198,14 +203,15 @@ export abstract class ChartProxy {
...
+        const rawChart = deproxy(chart);
+        const imageFileName = fileName || (rawChart.title ? rawChart.title.text : 'chart');
         const { width, height } = dimensions || {};
 
         AgChart.download(chart, { width, height, fileName: imageFileName, fileFormat });
     }
 
     public getChartImageDataURL(type?: string) {
-        return this.chart.scene.getDataURL(type);
+        return deproxy(this.chart).scene.getDataURL(type);
```
Comment (taktran) [inline enterprise-modules/charts/src/charts/chartComp/chartProxies/chartProxy.ts:214]: nitpick: I guess you could reuse `getChart()` here:

```suggestion
        return this.getChart().scene.getDataURL(type);
```

**Review by taktran — APPROVED** [review]
Looks good 👍🏼

File: charts-packages/ag-charts-community/src/chart/test/utils.ts [hunk 2]
```diff
@@ -38,6 +38,10 @@ process.env.FONTCONFIG_NAME = `${__dirname}/fonts.conf`;
 export const CANVAS_WIDTH = 800;
 export const CANVAS_HEIGHT = 600;
 
+export function deproxy(chartOrProxy: Chart | AgChartInstance): Chart {
```
Comment (alexanderby) [inline charts-packages/ag-charts-community/src/chart/test/utils.ts:41]: Screenshot 2022-11-22 at 11 45 14

**Review by alexanderby — APPROVED** [review]
Everything is great. Just curious if the transferrable resources are a part of a public API.

**Comment by alantreadway** [comment]
> Everything is great. Just curious if the transferrable resources are a part of a public API.

No, so the `AgChartInstance` contract is unchanged:
```
  destroy(): void;
```

Since `AgChartInstanceProxy` is the implementation users will now have, it strictly honours this contract. `Chart` however supports a super-set of this - but if invoked in-line with the `AgChartInstance` contract still honours that contract by returning `undefined` aka `void`.

**Comment by alantreadway** [comment]
@manapeirov's review is pending, but I want to minimise disruption to @taktran's change in terms of times he needs to resolve conflicts, so I'm going to merge and follow-up on any further feedback distinctly.

File: enterprise-modules/charts/src/charts/chartComp/gridChartComp.ts [hunk 3]
```diff
@@ -430,8 +433,8 @@ export class GridChartComp extends Component {
         return this.chartController.getChartId();
     }
 
-    public getUnderlyingChart(): any {
-        return this.chartProxy.getChart();
+    public getUnderlyingChart() {
+        return this.chartProxy.getChartRef();
```
Comment (manapeirov) [inline enterprise-modules/charts/src/charts/chartComp/gridChartComp.ts:437]: This is also a confusing name for this function, it's getting the chart instance proxy not chart?

**Pull request merged**